### ISSUE TITLE
Retry GET requests to GitHub if error is returned with a 200 status

### ISSUE THREAD

**Issue by synthead** [github/gh-gei#666]
## Description

On occasion, GET requests to GitHub will respond with an error that still has a 200 status code.  Here's what the verbose log looks like when this occurs:

```
[12:32] [INFO] Migration in progress (ID: <censored>). State: IN_PROGRESS. Waiting 10 seconds...
[12:32] [DEBUG] HTTP POST: https://api.github.com/graphql
[12:32] [DEBUG] HTTP BODY: {"query":"query($id: ID!) { node(id: $id) { ... on Migration { id, sourceUrl, migrationSource { name }, state, failureReason, repositoryName } } }","variables":{"id":"<censored>"}}
[12:32] [DEBUG] GITHUB REQUEST ID: 0401:086D:26F0E28:4F8CE9D:633325A8
[12:32] [DEBUG] RESPONSE (OK): {"errors":[{"message":"Something went wrong while executing your query. Please include `0401:086D:26F0E28:4F8CE9D:633325A8` when reporting this issue."}]}
[12:32] [ERROR] System.NullReferenceException: Object reference not set to an instance of an object.
   at OctoshiftCLI.GithubApi.GetMigration(String migrationId)
   at OctoshiftCLI.GithubEnterpriseImporter.Commands.MigrateRepoCommand.Invoke(MigrateRepoCommandArgs args)
   at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
   at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
```

Since a 200 status was returned, the CLI attempts to read data from the request, gets `null`, and raises a `System.NullReferenceException`.  Like https://github.com/github/gh-gei/issues/587, we should be safe to always retry all GET requests.  We'll want to also check if we're getting an "error" 200 in a response, and if so, consider it a failed request and perform a retry.

## Reproduction Steps

Reproduction steps are difficult since this is an intermittent issue.  However, the verbose log does include the entire JSON response, so we should be able to cater to its data when working on this issue.

**Comment by timrogers** [comment]
Thank you for creating this issue ❤️ It would be wonderful if we could get this implemented soon - anything we can do to avoid flakiness would be much appreciated 🙇🏻

**Comment by ArinGhazarian** [comment]
I just looked at the [CLI code](https://github.com/github/gh-gei/blob/11cd4edb9284edb6cf35db084f3bfeb2d080ec0b/src/Octoshift/GithubApi.cs#L405) and it looks like we’re already retrying! Which means that we are getting a 200 response back but with an error GQL payload! It probably won't hurt to retry on the CLI side for `GetMigration` method, but it would be a retry on a generic GQL error payload. It'd be also a good idea for the BE to return a more specific GQL error and also might worth doing some quick retries.

**Comment by timrogers** [comment]
I'd be in favour of trying this generic `200` GraphQL error, specifically for the `GetMigration` method.

**Comment by synthead** [comment]
https://github.com/github/github/pull/241010 just shipped to prod!  This change makes the Octoshift GraphQL APIs return `"type": "SERVICE_UNAVAILABLE"` in the error payload for failed requests.  We can consume this error message when solving for this issue to do informed retries so we're not just running blind :+1: 

I made sure in the PR that we only return these errors for requests that can be safely retried.  This includes:

- GraphQL queries that could not make a connection to Octoshift
- GraphQL queries that made a connection to Octoshift, but had a dropped connection while transferring data
- GraphQL mutations that could not make a connection to Octoshift

Note that I opted to still return a "something went wrong" failure for mutations that had their connections dropped partway because we cannot be confident that server-side transactions haven't been made.
